### ISSUE TITLE
NumPy backend configurable to use single-float precision

### ISSUE THREAD

**Issue by kratsg** [scikit-hep/pyhf#271]
# Description

When comparing many calculations between single-float and double-float precision backends, there will be a (sometimes non-negligible) deviation between the two because of this precision. This doesn't necessarily imply that the calculation is wrong, since it depends on the underlying precision of the floats (64-bit versus 32-bit). NumPy and Python by default use whatever is available on the system (e.g. 64-bit most of the time these days) while tensorflow and pytorch are using single-float precision (e.g. 32-bit).

We should make sure the numpy backend can be configured to use 32-bit precision through something like `pyhf.tensor.numpy_backend(use_single_floats=True)` as a flag that will pass in `dtype=np.float32` for all functions in the numpy backend.

# Relevant Issues and Pull Requests

- #251

**Comment by lukasheinrich** [comment]
this won't work since fitting at single precisiono often fails without exact gradients, instead we have added double precision for autodiff backends
